docs/robotframework.rst

```diff
@@ -200,6 +200,180 @@ In addition to browser interactions, the Salesforce Library also provides the fo
 * **Salesforce Update**: Updates a record using its type, ID, and field=value syntax
 * **SOQL Query**: Runs a SOQL query and returns a REST API result dictionary
 
+PageObjects Library
+===================
+
+The **PageObjects** library provides support for page objects,
+robotframework-style. Even though robot is a keyword driven framework,
+we've implemented a way to dynamically load in keywords that are
+unique to a page or an object on the page.
+
+With this library, you can define classes which represent page
+objects. Each class has keywords that are unique to a page or a
+component. These classes can be imported on demand only for tests
+which use these pages or components.
+
+The pageobject Decorator
+------------------------
+
+Page objects are normal python classes which use the :code:`pageobject`
+decorator provided by cumulusci. Unlike traditional robot framework
+keyword libraries, you may define multiple sets of keywords in a
+single file.
+
+For example, the following example shows the definition of a page
+object for the listing page of the Contact object. It adds a new
+keyword named :code:`Click contact link`:
+
+.. code-block:: python
+
+   from cumulusci.robotframework.pageobjects import pageobject, ListingPage
+
+   @pageobject(page_type="Listing", object_name="Contact")
+   class ContactListingPage(ListingPage):
+
+       def click_contact_link(self, name):
+           self.selenium.click_link('xpath://a[@title="{}"]'.format(name))
+           self.salesforce.wait_until_loading_is_complete()
+
+The :code:`pageobject` decorator takes two arguments: :code:`page_type` and
+:code:`object_name`. These two arguments are used to identify the page
+object (eg: :code:`Go To Page  Listing  Contact`). The values can be
+any arbitrary string, but ordinarily should represent standard page
+types ("Listing", "Detail", "Home"), and standard object names.
```

Comment:
The developer name or the user-facing label for the object? If the former, we need to talk about how to make this work both when objects have a namespace prefix and when they don't.

Reply:
Yes, I haven't given namespaces any consideration. I haven't yet taken the time to grok how they work.

Reply:
It's not super complicated (though dealing with the implications can be), and I think it's worth grokking before we introduce a framework that assumes objects are always named the same thing.

Standard objects like Contact always have the same name.

Custom objects which are not in a managed package or a namespaced scratch org have a name with a `__c` suffix like MyObject__c. So far no problem.

Custom objects which are in a managed package add a prefix using the package's globally unique namespace. For example npsp__Address__c is the developer name of the Address object from NPSP when it is installed as a managed package. Objects also get this namespace prefix if they are in an unmanaged package being developed in a scratch org that has its own namespace.

Since our products are typically developed as unmanaged packages and then released as managed packages, that means that the Address object would be called Address__c during development but npsp__Address__c once it is released (or in development in a namespaced scratch org).

Since we want to run our robot tests both during development and with managed packages for regression, we need to figure out how to handle this.

Reply:
We probably need to discuss this over chat or a video call, but let me get a couple of basic questions out of the way.

At the time that a developer needs to write a keyword, do they know what the namespace is / will be? Could the solution be as simple as defining the namespace in the class or decorator, and then at runtime we do something like this:

    @pageobject(page_type="Listing", object_name="MyObject__c", namespace="npsp")
    def whatever(ListingPage):
        ...
        object_name = self.namespace + self.object_name if self.is_managed() else self.object_name
        url = ".../lightning/o/{}/list".format(object_name)"
        ...

Or is it a bit more complicated than that? Is it possible to write a `self.is_managed` function or property that can use introspection to get the answer (and/or a `self.get_namespace_prefix()` function)?

Reply:
poking about through the NPSP code I see that it does `self.cumulusci.sf.describe()` and then does a bit of string manipulation to get the namespace prefix. Would that be the foundation of `self.is_managed()` or `self.get_namespace_prefix()`?

Reply:
That sounds like the right direction, but we might not even need to add the `namespace` argument because it's typically only objects from the current repository that would vary between namespaced and un-namespaced, so we can look up the current project's namespace from cumulusci.yml. (Underlying packages would be installed as managed packages and always have their namespace.)

There's already a keyword in the CumulusCI library, `Get Namespace Prefix` to get the namespace prefix for a specified package (defaulting to the package for the current repository) or an empty string if there's no prefix.

`object_name` could become a property that encapsulates the logic for adding or not-adding the prefix (and memoizes it).

Reply:
Will namespaces always only apply to custom objects? That is, if I set `object_name` to `Contact`, I can determine it will never have a namespace associated with it, but if the object is named `MyObject__c`, then I need to do the dance to maybe prepend the namespace?

I (think I) want to support the ability for a teams to create page objects for standard objects, until we reach the point that we have full coverage. For example, NPSP or GEM or whomever might create a `Contact` page object or `Opportunity` page object to get them over a hump, which we can eventually fold into the core. In such a case, we wouldn't want to use the namespace logic for those page objects.

So, the logic for building a URL might look like:

```
    if object_name.endswith("__c") and is_managed():
        prepend the namespace
    else
        don't prepend the namespace
```

Reply:
That's right that it's only for custom objects. It's also only for custom objects that don't already have a different namespace. I think the logic could be something like this:

    @property
    def object_name(self):
        object_name = self._object_name # assuming we change the decorator to set _object_name instead of object_name
        parts = object_name.split("__")
        if len(parts) == 2 and parts[-1] == "c":  # the length check is to skip objects from a different namespace like foobar__otherpackageobject__c
            object_name = self.cumulusci.get_namespace_prefix() + object_name # get_namespace_prefix already takes care of returning an actual prefix or an empty string depending on whether the package is managed
        return object_name